Cubical/Categories/Site/Sheafification/Base.agda

```diff
@@ -0,0 +1,126 @@
+-- using a quotient inductive type (QIT).
+
+-- This is inspired by the construction of the sheafification (for finite coverings) in:
+-- * E. Palmgren, S.J. Vickers, "Partial Horn logic and cartesian categories"
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Structure
+open import Cubical.Foundations.Function using (_∘_)
+
+open import Cubical.HITs.PropositionalTruncation using (∣_∣₁)
+
+open import Cubical.Data.Sigma
+
+open import Cubical.Functions.Surjection
+open import Cubical.Functions.Embedding
+
+open import Cubical.Categories.Category
+open import Cubical.Categories.Presheaf
+open import Cubical.Categories.Functor
+
+open import Cubical.Categories.Site.Cover
+open import Cubical.Categories.Site.Coverage
+open import Cubical.Categories.Site.Sheaf
+
+
+module Sheafification
+  {ℓ ℓ' ℓcov ℓpat : Level}
+  {C : Category ℓ ℓ'}
+  (J : Coverage C ℓcov ℓpat)
+  {ℓP : Level}
+  (P : Presheaf C ℓP)
+  where
+
+  open Category C hiding (_∘_)
+  open Coverage J
+
+  data ⟨F⟅_⟆⟩ : ob → Type (ℓ-max ℓ (ℓ-max ℓ' (ℓ-max ℓcov (ℓ-max ℓpat ℓP)))) where
+
+    trunc : {c : ob} → isSet ⟨F⟅ c ⟆⟩
+
+    restrict : {c d : ob} → (f : Hom[ d , c ]) → ⟨F⟅ c ⟆⟩ → ⟨F⟅ d ⟆⟩
+
+    restrictId :
+      {c : ob} →
+      (x : ⟨F⟅ c ⟆⟩) →
+      restrict id x ≡ x
+    restrictRestrict :
+      {c d e : ob} →
+      (f : Hom[ d , c ]) →
+      (g : Hom[ e , d ]) →
+      (x : ⟨F⟅ c ⟆⟩) →
+      restrict (g ⋆ f) x ≡ restrict g (restrict f x)
+
+    η⟦⟧ : {c : ob} → (x : ⟨ P ⟅ c ⟆ ⟩) → ⟨F⟅ c ⟆⟩
+
+    ηNatural :
+      {c d : ob} →
+      (f : Hom[ d , c ]) →
+      (x : ⟨ P ⟅ c ⟆ ⟩) →
+      η⟦⟧ ((P ⟪ f ⟫) x) ≡ restrict f (η⟦⟧ x)
+
+    sep :
+      {c : ob} →
+      (cover : ⟨ covers c ⟩) →
+      let cov = str (covers c) cover in
+      (x y : ⟨F⟅ c ⟆⟩) →
+      (i~j :
+        (patch : ⟨ cov ⟩) →
+        restrict (patchArr cov patch) x ≡ restrict (patchArr cov patch) y) →
+      x ≡ y
+
+    amalgamate :
+      let
+      -- Is there any way to make this definition now and reuse it later?
+      F : Presheaf C _
+      F = record
+        { F-ob = λ c → ⟨F⟅ c ⟆⟩ , trunc
+        ; F-hom = restrict
+        ; F-id = funExt restrictId
+        ; F-seq = λ f g → funExt (restrictRestrict f g)
+        }
+      in
+      {c : ob} →
+      (cover : ⟨ covers c ⟩) →
+      let cov = str (covers c) cover in
+      (fam : CompatibleFamily F cov) →
+      ⟨F⟅ c ⟆⟩
+    restrictAmalgamate :
+      let
+      F : Presheaf C _
+      F = record
+        { F-ob = λ c → ⟨F⟅ c ⟆⟩ , trunc
+        ; F-hom = restrict
+        ; F-id = funExt restrictId
+        ; F-seq = λ f g → funExt (restrictRestrict f g)
+        }
+      in
+      {c : ob} →
+      (cover : ⟨ covers c ⟩) →
+      let cov = str (covers c) cover in
+      (fam : CompatibleFamily F cov) →
+      (patch : ⟨ cov ⟩) →
+      restrict (patchArr cov patch) (amalgamate cover fam) ≡ fst fam patch
+
+  F : Presheaf C (ℓ-max (ℓ-max (ℓ-max (ℓ-max ℓ ℓ') ℓcov) ℓpat) ℓP)
```

Comment:
Should this not have a more descriptive name, like `sheafification`?

Reply:
The exposed name is `Sheafification.F`. Do you think it should still be called `sheafification` inside the named module `Sheafification`?

Reply:
I think I actually went for a named module precisely to give users the option to open it with arguments (either just `J` or both `J` and `P`). So let's go with `sheafification` and `isSheafSheafification`, right?

Reply:
Side remark: I can think of the following notations for sheafification in the literature, which seem not very helpful either.

- $a$ (for "associated sheaf functor") [MacLane, Moerdijk: Sheaves in Geometry an Logic]
- $i^*$ (because it is the inverse image part of the geometric morphism from sheaves to presheaves)

Reply:
> it would probably be best to make `F⟅_⟆` private.

That also makes all the constructors of the HIT private, unfortunately. At least I can't find a way to mark only the name `⟨F⟅_⟆⟩` as private.

So I guess I should rename it to `⟨sheafification⟅_⟆⟩`.

Reply:
I replaced all `F`s now (that stood for a sheafification). Good like that?

Reply:
Yeah, looks good to me, but it is of course a purely stylistic matter. There are instances of both in the library, named modules with short names, like `f` for a certain function, and named modules with terms that have essentially the same name as the module (intended to opened). I don't want to advocate too heavily for one of the options, but if everyone is happy now, we can leave things as they are, I would say.